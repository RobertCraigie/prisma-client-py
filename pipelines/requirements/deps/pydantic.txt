pydantic==2.3.0
