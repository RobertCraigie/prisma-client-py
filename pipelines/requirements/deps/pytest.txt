pytest==7.4.4
