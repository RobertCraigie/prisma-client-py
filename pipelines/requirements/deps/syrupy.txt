syrupy==4.6.4
