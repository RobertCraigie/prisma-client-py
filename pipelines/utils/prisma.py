from __future__ import annotations

import nox


def generate(
    session: nox.Session,
    *,
    schema: str | None = 'tests/data/schema.prisma',
    clean: bool = True,
) -> None:
    if clean:
        session.run('python', '-m', 'prisma_cleanup')

    if schema:
        args = (f'--schema={schema}',)
    else:
        args = ()

    session.run('prisma', 'generate', *args)
