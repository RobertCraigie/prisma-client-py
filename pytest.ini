[pytest]
markers =
  persist_data: Signifies that data should not be cleared between test runs.
  prisma: This marker must be used to enable model-based access.

# asyncio fixtures and tests must be explicitly marked
asyncio_mode = strict

# NOTE: this is required to stop pytest from loading conftest.py files
# for our integration tests, as pytest eagerly loads conftests backwards
# we cannot use a hook to ignore them as the integration conftest files
# are loaded before ours, we might be able to fix this by installing
# our own pytest plugin?
norecursedirs =
  integrations/*
  tests/integrations/postgresql/*
