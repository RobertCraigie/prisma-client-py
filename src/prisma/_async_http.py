from typing import Any

import httpx

from ._types import Method
from .http_abstract import AbstractResponse, AbstractHTTP


__all__ = ('HTTP', 'Response', 'client')


class HTTP(AbstractHTTP[httpx.AsyncClient, httpx.Response]):
    # pylint: disable=invalid-overridden-method,attribute-defined-outside-init

    session: httpx.AsyncClient

    async def download(self, url: str, dest: str) -> None:
        # pyright thinks that stream() doesn't return a context manager
        # due to how httpx handles compatability imports
        # https://github.com/encode/httpx/discussions/1829
        async with self.session.stream(
            'GET', url, timeout=None
        ) as resp:  # pyright: reportGeneralTypeIssues=false
            with open(dest, 'wb') as fd:
                async for chunk in resp.aiter_bytes():
                    fd.write(chunk)

    async def request(self, method: Method, url: str, **kwargs: Any) -> 'Response':
        return Response(await self.session.request(method, url, **kwargs))

    def open(self) -> None:
        self.session = httpx.AsyncClient()

    async def close(self) -> None:
        if not self.closed:
            await self.session.aclose()

            # mypy doesn't like us assigning None as the type of
            # session is not optional, however the argument that
            # the setter takes is optional, so this is fine
            self.session = None  # type: ignore[assignment]


client: HTTP = HTTP()


class Response(AbstractResponse[httpx.Response]):
    # pylint: disable=invalid-overridden-method

    @property
    def status(self) -> int:
        return self.original.status_code

    async def json(self, **kwargs: Any) -> Any:
        return self.original.json(**kwargs)

    async def text(self, **kwargs: Any) -> Any:
        return self.original.content.decode(**kwargs)
