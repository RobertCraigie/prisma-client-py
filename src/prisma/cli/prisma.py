import os
import sys
import logging
import subprocess
from textwrap import indent
from typing import List, Optional, Dict

import click

from .. import binaries


log: logging.Logger = logging.getLogger(__name__)


def run(
    args: List[str],
    check: bool = False,
    env: Optional[Dict[str, str]] = None,
) -> int:
    directory = binaries.ensure_cached()
    path = directory.joinpath(binaries.PRISMA_CLI_NAME)
    if not path.exists():
        raise RuntimeError(
            f'The Prisma CLI is not downloaded, expected {path} to exist.'
        )

    log.debug('Using Prisma CLI at %s', path)
    log.debug('Running prisma command with args: %s', args)

    default_env = {
        **os.environ,
        'PRISMA_HIDE_UPDATE_MESSAGE': 'true',
        'PRISMA_CLI_QUERY_ENGINE_TYPE': 'binary',
    }
    env = {**default_env, **env} if env is not None else default_env
    # ensure the client uses our engine binaries
    for engine in binaries.ENGINES:
        env[engine.env] = str(engine.path.absolute())

    if args and args[0] == 'studio':
        click.echo(
            click.style(
                'ERROR: Prisma Studio does not work natively with Prisma Client Python',
                fg='red',
            ),
        )
        click.echo(
            '\nThere are two other possible ways to use Prisma Studio:\n'
        )
        click.echo(
            click.style('1. Download the Prisma Studio app\n', bold=True)
        )
        click.echo(
            indent(
                'Prisma Studio can be downloaded from: '
                + click.style(
                    'https://github.com/prisma/studio/releases', underline=True
                ),
                ' ' * 3,
            )
        )
        click.echo(
            click.style('\n2. Use the Node based Prisma CLI:\n', bold=True)
        )
        click.echo(
            indent(
                'If you have Node / NPX installed you can launch Prisma Studio by running the command: ',
                ' ' * 3,
            ),
            nl=False,
        )
        click.echo(click.style('npx prisma studio', bold=True))
        return 1

    process = subprocess.run(
        [str(path.absolute()), *args],
        env=env,
        check=check,
        stdout=sys.stdout,
        stderr=sys.stderr,
    )

    if args and args[0] in {'--help', '-h'}:
        prefix = ' '
        click.echo(click.style(prefix + 'Python Commands\n', bold=True))
        click.echo(
            prefix
            + 'For Prisma Client Python commands see '
            + click.style('prisma py --help', bold=True)
        )

    return process.returncode
