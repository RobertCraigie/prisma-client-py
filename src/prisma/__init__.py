# -*- coding: utf-8 -*-

__title__ = 'prisma'
__author__ = 'RobertCraigie'
__license__ = 'APACHE'
__copyright__ = 'Copyright 2020-2023 RobertCraigie'
__version__ = '0.13.0'

from typing import TYPE_CHECKING

from . import errors as errors
from .utils import setup_logging
from ._types import PrismaMethod as PrismaMethod
from ._config import config as config
from ._metrics import (
    Metric as Metric,
    Metrics as Metrics,
    MetricHistogram as MetricHistogram,
)
from .validator import *

# the import ordering is important here because
# we rely on the fact that `prisma/client.py` is the
# first piece of generated code that is loaded. This is
# especially useful for `python -m prisma_cleanup`
try:
    from .client import *  # noqa: I001, TID251
    from .fields import *  # noqa: TID251
    from . import (
        bases as bases,  # noqa: TID251
        types as types,  # noqa: TID251
        models as models,  # noqa: TID251
        partials as partials,  # noqa: TID251
    )
except ModuleNotFoundError:
    # code has not been generated yet
    # TODO: this could swallow unexpected errors

    # we only define this magic function when the client has not
    # been generated for performance reasons and we hide it from
    # type checkers so that they don't incidentally disable type
    # checking due to the dynamic nature of the `__getattr__`
    # magic dunder method.
    if not TYPE_CHECKING:

        def __getattr__(name: str):
            try:
                return globals()[name]
            except KeyError as err:
                # TODO: support checking for 'models' here too
                if name in {'Prisma', 'Client'}:
                    # TODO: remove this frame from the stack trace
                    raise RuntimeError(
                        "The Client hasn't been generated yet, "
                        'you must run `prisma generate` before you can use the client.\n'
                        'See https://prisma-client-py.readthedocs.io/en/stable/reference/troubleshooting/#client-has-not-been-generated-yet'
                    ) from None

                # leaves handling of this potential error to Python as per PEP 562
                raise AttributeError() from err


setup_logging()
