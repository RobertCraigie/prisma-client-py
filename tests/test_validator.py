from typing import Any, cast

import pytest
from pydantic import Extra, BaseConfig, ValidationError, annotated_types
from syrupy.assertion import SnapshotAssertion

import prisma
from prisma import Client, validate, types


class Foo:
    pass


def test_valid() -> None:
    """Basic usage with correct data"""
    validated = validate(types.IntFilter, {'equals': 1})
    assert validated == {'equals': 1}  # type: ignore[comparison-overlap]

    validated = validate(types.IntFilter, {'equals': '1'})
    assert validated == {'equals': 1}  # type: ignore[comparison-overlap]


def test_disallows_non_typeddict_type(snapshot: SnapshotAssertion) -> None:
    """Validating against non TypedDict types throws an error"""
    with pytest.raises(TypeError) as exc:
        validate(Foo, None)

    assert str(exc.value) == snapshot

    with pytest.raises(TypeError) as exc:
        validate(dict, None)

    assert str(exc.value) == snapshot


def test_non_dictionary_values(snapshot: SnapshotAssertion) -> None:
    """Validating a non-dictionary value throws an error"""
    with pytest.raises(ValidationError) as exc:
        validate(types.UserCreateInput, None)

    assert str(exc.value) == snapshot


def test_recursive(snapshot: SnapshotAssertion) -> None:
    """Validating recursive types works as expected"""
    with pytest.raises(ValidationError) as exc:
        validate(types.FloatFilter, {'not': {'not': {'not': 'a'}}})

    assert str(exc.value) == snapshot

    validated = validate(types.FloatFilter, {'not': {'not': {'not': '193.4'}}})
    assert validated == {'not': {'not': {'not': 193.4}}}  # type: ignore[comparison-overlap]


def test_missing_values(snapshot: SnapshotAssertion) -> None:
    """TypedDict with required keys is correctly validated"""
    with pytest.raises(ValidationError) as exc:
        validate(types.PostCreateInput, {})

    assert str(exc.value) == snapshot


def test_optional_values() -> None:
    """Fields that can be None are still included in the validated data"""
    validated = validate(
        types.PostCreateInput, dict(title='My Title', published=True)
    )
    assert validated == {'title': 'My Title', 'published': True}  # type: ignore[comparison-overlap]

    validated = validate(
        type=types.PostCreateInput,
        data=dict(title='My Title', published=True, desc=None),
    )
    assert validated == {  # type: ignore[comparison-overlap]
        'title': 'My Title',
        'published': True,
        'desc': None,
    }


def test_disallows_extra_values(snapshot: SnapshotAssertion) -> None:
    """Fields that are not present in the TypedDict are not allowed"""
    with pytest.raises(ValidationError) as exc:
        validate(types.PostCreateInput, {'foo': 'bar'})

    assert str(exc.value) == snapshot


class AllowExtraConfig(BaseConfig):
    extra: Extra = Extra.allow


class DisallowExtraConfig(BaseConfig):
    extra: Extra = Extra.forbid


def test_patch_respects_arguments(snapshot: SnapshotAssertion) -> None:
    """Ensure that creating a BaseModel from a TypedDict does not cache
    the BaseModel irrespective of the given __config__
    """
    Model = cast(
        Any,
        annotated_types.create_model_from_typeddict(
            types.UserCreateInput, __config__=AllowExtraConfig  # type: ignore
        ),
    )
    with pytest.raises(ValidationError) as exc:
        Model()

    assert str(exc.value) == snapshot

    model = Model(name='Robert')
    assert model.name == 'Robert'

    model = Model(name='Tegan', foo='bar')
    assert model.name == 'Tegan'
    assert model.foo == 'bar'

    Model = cast(
        Any,
        annotated_types.create_model_from_typeddict(
            types.UserCreateInput, __config__=DisallowExtraConfig  # type: ignore
        ),
    )
    with pytest.raises(ValidationError) as exc:
        Model(name='Robert', foo='bar')

    assert str(exc.value) == snapshot


@pytest.mark.asyncio
async def test_validate_arguments(
    client: Client,
    snapshot: SnapshotAssertion,
) -> None:
    """Ensure that passing incorrect arguments to an action method will raise a ValidationError"""
    with pytest.raises(ValidationError) as exc:
        await client.user.create({})  # type: ignore

    assert str(exc.value) == snapshot


@pytest.mark.asyncio
async def test_disable_validation(client: Client) -> None:
    """Ensure that query argument validation can be skipped"""
    with prisma.disable_validation():
        with pytest.raises(prisma.errors.MissingRequiredValueError):
            await client.user.create({})  # type: ignore


@pytest.mark.asyncio
async def test_nested_disable_validation(client: Client) -> None:
    """Ensure that nesting two disable_validation calls
    does not reset state prematurely
    """
    with prisma.disable_validation():
        with prisma.disable_validation():
            with pytest.raises(prisma.errors.MissingRequiredValueError):
                await client.user.create({})  # type: ignore

        with pytest.raises(prisma.errors.MissingRequiredValueError):
            await client.user.create({})  # type: ignore
