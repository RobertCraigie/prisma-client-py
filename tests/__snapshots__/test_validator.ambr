# name: test_disallows_extra_values
  '
  3 validation errors for PostCreateInput
  title
    field required (type=value_error.missing)
  published
    field required (type=value_error.missing)
  foo
    extra fields not permitted (type=value_error.extra)
  '
---
# name: test_disallows_non_typeddict_type
  "Only TypedDict types are supported, got: <class 'tests.test_validator.Foo'> instead."
---
# name: test_disallows_non_typeddict_type.1
  "Only TypedDict types are supported, got: <class 'dict'> instead."
---
# name: test_missing_values
  '
  2 validation errors for PostCreateInput
  title
    field required (type=value_error.missing)
  published
    field required (type=value_error.missing)
  '
---
# name: test_non_dictionary_values
  '
  1 validation error for UserCreateInput
  __root__
    UserCreateInput expected dict not NoneType (type=type_error)
  '
---
# name: test_patch_respects_arguments
  '
  1 validation error for UserCreateInput
  name
    field required (type=value_error.missing)
  '
---
# name: test_patch_respects_arguments.1
  '
  1 validation error for UserCreateInput
  foo
    extra fields not permitted (type=value_error.extra)
  '
---
# name: test_recursive
  '
  4 validation errors for FloatFilter
  not
    value is not a valid float (type=type_error.float)
  not -> not
    value is not a valid float (type=type_error.float)
  not -> not -> not
    value is not a valid float (type=type_error.float)
  not -> not -> not -> __root__
    FloatFilterRecursive3 expected dict not str (type=type_error)
  '
---
# name: test_validate_arguments
  '
  1 validation error for Create
  data -> name
    field required (type=value_error.missing)
  '
---
